internal/service/service.go

```diff
@@ -33,9 +36,32 @@ func (s *URLService) Shorten(baseURL string, originalURL string) (string, error)
 		return "", fmt.Errorf("url is empty: %w ", ErrInvalidURL)
 	}
 
-	id, err := s.repo.Save(originalURL)
+	const maxAttempts = 10 // Максимальное количество попыток
+	var id string
+	attempts := 0
 
-	if err != nil {
+	for attempts < maxAttempts {
+		genID := s.idGenerator.GenerateID()
+		_, err := s.repo.Find(genID)
+
+		if errors.Is(err, repository.ErrURLNotFound) {
+			id = genID
+			break
+		}
+
+		if err != nil {
+			log.Println("Error finding id:", err)
+			return "", fmt.Errorf("failed to find id %s: %w", id, err)
+		}
+		attempts++
+	}
+
+	if id == "" {
+		log.Printf("Number of attempts exceeded: %s", originalURL)
+		return "", ErrInternalServer
```

Comment:
В ошибку можно добавить контекст, чтобы залогировать выше.

```suggestion
		return "",  fmt.Errorf("%w: number of attempts exceeded: %s", ErrInternalServer, originalURL)
```